File: lib/metrics.ts
function dist(a: { x: number; y: number }, b: { x: number; y: number }): number {
  return Math.sqrt(Math.pow(a.x - b.x, 2) + Math.pow(a.y - b.y, 2));
}

function px(lm: { x: number; y: number }, w: number, h: number): [number, number] {
  return [lm.x * w, lm.y * h];
}

export function computeMetrics(landmarks: any[], w: number, h: number): Record<string, number> {
  const leftEye = landmarks[33];
  const rightEye = landmarks[263];

  const [lx, ly] = px(leftEye, w, h);
  const [rx, ry] = px(rightEye, w, h);

  const headTilt = (Math.atan2(ry - ly, rx - lx) * 180) / Math.PI;

  // Calculate eye openness more accurately using both eyes
  const leftEyeTop = landmarks[159];
  const leftEyeBottom = landmarks[145];
  const leftEyeHeight = dist(leftEyeTop, leftEyeBottom);
  
  const rightEyeTop = landmarks[386];
  const rightEyeBottom = landmarks[374];
  const rightEyeHeight = dist(rightEyeTop, rightEyeBottom);
  
  const leftEyeLeftCorner = landmarks[33];
  const leftEyeRightCorner = landmarks[133];
  const leftEyeWidth = dist(leftEyeLeftCorner, leftEyeRightCorner);
  
  const rightEyeLeftCorner = landmarks[362];
  const rightEyeRightCorner = landmarks[263];
  const rightEyeWidth = dist(rightEyeLeftCorner, rightEyeRightCorner);
  
  // Calculate openness ratio for both eyes and average
  const leftEyeOpenness = leftEyeWidth > 0 ? leftEyeHeight / leftEyeWidth : 0;
  const rightEyeOpenness = rightEyeWidth > 0 ? rightEyeHeight / rightEyeWidth : 0;
  const eyeOpenness = (leftEyeOpenness + rightEyeOpenness) / 2;

  const lipLeft = landmarks[61];
  const lipRight = landmarks[291];
  const lipTop = landmarks[13];

  const leftDist = dist(lipLeft, lipTop);
  const rightDist = dist(lipRight, lipTop);

  const smileSymmetry =
    leftDist + rightDist === 0
      ? 1.0
      : 1.0 - Math.abs(leftDist - rightDist) / (leftDist + rightDist);

  const browLeft = landmarks[70];
  const browRight = landmarks[300];
  const eyeCenter = landmarks[168];

  const leftBrowDist = dist(browLeft, eyeCenter);
  const rightBrowDist = dist(browRight, eyeCenter);

  const browSymmetry =
    leftBrowDist + rightBrowDist === 0
      ? 1.0
      : 1.0 - Math.abs(leftBrowDist - rightBrowDist) / (leftBrowDist + rightBrowDist);

  const mouthTop = landmarks[13];
  const mouthBottom = landmarks[14];
  const mouthOpen = dist(mouthTop, mouthBottom);

  const chin = landmarks[152];
  const forehead = landmarks[10];
  const faceH = dist(chin, forehead);
  const mouthOpenness = faceH > 0 ? mouthOpen / faceH : 0;

  // Tension index: measures asymmetry between left and right eye opening
  // More accurate calculation using normalized differences
  // Reuse leftEyeHeight and rightEyeHeight already calculated above
  const avgEyeOpen = (leftEyeHeight + rightEyeHeight) / 2;
  
  // Calculate relative asymmetry (difference / average)
  const eyeAsymmetry = avgEyeOpen > 0 ? Math.abs(leftEyeHeight - rightEyeHeight) / avgEyeOpen : 0;
  
  // Also check for facial tension indicators: brow position
  // Reuse browLeft and browRight already declared above
  const browAsymmetry = Math.abs(browLeft.y - browRight.y);
  
  // Normalize by face height and combine indicators
  // Reuse faceH already calculated above
  const normalizedBrowAsymmetry = faceH > 0 ? browAsymmetry / faceH : 0;
  
  // Combine eye and brow asymmetry, normalize to 0-1
  // Higher values = more tension
  const tensionIndex = Math.min(1, (eyeAsymmetry * 0.6 + normalizedBrowAsymmetry * 10 * 0.4));

  const confidenceIndex = Math.max(
    0.0,
    Math.min(
      1.0,
      smileSymmetry * 0.4 + eyeOpenness * 0.3 + (1 - tensionIndex) * 0.3
    )
  );

  return {
    head_tilt: headTilt,
    eye_openness: eyeOpenness,
    smile_symmetry: smileSymmetry,
    brow_symmetry: browSymmetry,
    mouth_openness: mouthOpenness,
    tension_index: tensionIndex,
    confidence_index: confidenceIndex,
  };
}

File: metrics.py
import math

def dist(a, b):
    return math.sqrt((a.x - b.x) ** 2 + (a.y - b.y) ** 2)

def compute_metrics(landmarks, w, h):
    def px(lm):
        return (lm.x * w, lm.y * h)

        left_eye = landmarks[33]
        right_eye = landmarks[263]

        lx, ly = px(left_eye)
        rx, ry = px(right_eye)

        head_tilt = math.degrees(math.atan2(ry - ly, rx - lx))

        eye_top = landmarks[159]
        eye_bottom = landmarks[145]
        topb = dist(eye_top, eye_bottom)
        
        eye_left_corner = landmarks[133]
        eye_right_corner = landmarks[33]
        eye_width = dist(eye_left_corner, eye_right_corner)
        eye_openness = (topb / eye_width) if eye_width > 0 else 0

        lip_left = landmarks[61]
        lip_right = landmarks[291]
        lip_top = landmarks[13]

        left_dist = dist(lip_left, lip_top)
        right_dist = dist(lip_right, lip_top)

        if left_dist + right_dist == 0:
            smile_symmetry = 1.0
        else:
            smile_symmetry = 1.0 - abs(left_dist - right_dist) / (left_dist + right_dist)

        brow_left = landmarks[70]
        brow_right = landmarks[300]
        eye_center = landmarks[168]

        left_brow_dist = dist(brow_left, eye_center)
        right_brow_dist = dist(brow_right, eye_center)

        if left_brow_dist + right_brow_dist == 0:
            brow_symmetry = 1.0
        else:
            brow_symmetry = 1.0 - abs(left_brow_dist - right_brow_dist) / (left_brow_dist + right_brow_dist)
        
        mouth_top = landmarks[13]
        mouth_bottom = landmarks[14]
        mouth_open = dist(mouth_top, mouth_bottom)

        chin = landmarks[152]
        forehead = landmarks[10]
        face_h = dist(chin, forehead)
        mouth_openness = (mouth_open / face_h) if face_h > 0 else 0

        tension = abs(dist(landmarks[159], landmarks[145]) - dist(landmarks[386], landmarks[374]))
        tension_index = min(max(tension * 5, 0), 1)

        confidence_index = ((smile_symmetry * 0.4) + (eye_openness * 0.3) + (1 - tension_index) * 0.3)
        confidence_index = max(0.0, min(1.0, confidence_index))

        return {
            "head_tilt": head_tilt,
            "eye_openness": eye_openness,
            "smile_symmetry": smile_symmetry,
            "brow_symmetry": brow_symmetry,
            "mouth_openness": mouth_openness,
            "tension_index": tension_index,
            "confidence_index": confidence_index
        }